lms/models/h_group.py

```diff
@@ -1,59 +1,36 @@
-from typing import NamedTuple
+import sqlalchemy as sa
 
-from lms.models._hashed_id import hashed_id
+from lms.db import BASE, TimestampedModelMixin
 
 MAX_GROUP_NAME_LENGTH = 25
 
 
-class HGroup(NamedTuple):
-    name: str
-    authority_provided_id: str
-    type: str
+class HGroup(TimestampedModelMixin, BASE):
+    __tablename__ = "h_group"
+    __table_args__ = (sa.UniqueConstraint("authority_provided_id"),)
+
+    id = sa.Column(sa.Integer(), autoincrement=True, primary_key=True)
+
+    # Full name of the group in the LMS
+    _name = sa.Column(
+        sa.String(),
+        nullable=False,
+    )
+    authority_provided_id = sa.Column(sa.UnicodeText(), nullable=False)
+
+    type = sa.Column(
+        sa.String(),
+        nullable=False,
+    )
 
     def groupid(self, authority):
         return f"group:{self.authority_provided_id}@{authority}"
 
-    @classmethod
-    def course_group(cls, course_name, tool_consumer_instance_guid, context_id):
-        """
-        Create an HGroup for a course.
-
-        :param course_name: The name of the course
-        :param tool_consumer_instance_guid: Tool consumer GUID
-        :param context_id: Course id
-        """
-        return HGroup(
-            cls._name(course_name),
-            hashed_id(tool_consumer_instance_guid, context_id),
-            type="course_group",
-        )
-
-    @classmethod
-    def section_group(
-        cls, section_name, tool_consumer_instance_guid, context_id, section_id
-    ):
-        """
-        Create an HGroup for a course section.
-
-        :param section_name: The name of the section
-        :param tool_consumer_instance_guid: Tool consumer GUID
-        :param context_id: Course id the section is a part of
-        :param section_id: A section id for a section group
-        """
-        return HGroup(
-            cls._name(section_name),
-            hashed_id(tool_consumer_instance_guid, context_id, section_id),
-            type="section_group",
-        )
-
-    @classmethod
-    def _name(cls, name):
+    @property
+    def name(self):
```

Comment:
I think there might be something not quite right about the `HGroup` model.

We're storing the name of the LMS course, section or group that the `HGroup` represents in an `h_group._name` column on the DB, then we have this `name` property that generates the `HGroup`'s actual name from the course, section or group name.

I think we may need to start recording LMS courses, Canvas sections and Canvas groups in the DB first. Courses are actually already done (https://github.com/hypothesis/lms/blob/74bdda7ae901cdd75139574ca5d035c1a7912474/lms/models/course.py) but Canvas sections and Canvas groups aren't.

Each course, Canvas section and Canvas group would be related to exactly one `HGroup` by a  non-nullable foreign key.

Then instead of `_name`, the `name()` property would generate the h group name by looking up the name of this `HGroup`'s related LTI course, Canvas section or Canvas group.

I'm not sure how exactly to implement this foreign key relation though. We want each group to be related to _exactly one_ LTI course, Canvas section or Canvas group in a `1:1` relation. But courses, sections and groups will probably be in three separate DB tables. I'm sure it can be done somehow